### DISCUSSION TITLE
Variability of words / phrases that are matching with specific label

### DISCUSSION THREAD

**Discussion by aleksandar-devedzic** [explosion/spaCy#7570] [Help: Best practices]
I want to make a NER model that will recognise street addresses. 
That kind of data have a big variability, for example.

[8237 Monroe Drive
Bountiful, UT 84010
16 Dogwood Ave.
Grand Island, NE 68801
634 Poplar Ave.
Elyria, OH 44035
79 Sheffield Dr.
Cranford, NJ 07016
97 Colonial Dr.
Salem, MA 01970
211 Oakland St.
Yakima, WA 98908]
...
+much more names (that are depending on country etc.)
...

And if I want to train my model for street recognition:
```
training_data = [('send to: Aargauerstrasse 8005', {'entities': [(9, 28, 'ADDRESS')]}), 
                ('send to: Abeggweg 8057', {'entities': [(9, 21, 'ADDRESS')]}), 
                ('send to: Abendweg 8038', {'entities': [(9, 21, 'ADDRESS')]}), 
                ('send to: Ackermannstrasse 8044', {'entities': [(9, 29, 'ADDRESS')]}), 
                ('send to: Aehrenweg 8050', {'entities': [(9, 22, 'ADDRESS')]}), 
                ('send to: Aemmerliweg 8050', {'entities': [(9, 24, 'ADDRESS')]}), 
                ('send to: Albisgütliweg 8045', {'entities': [(9, 26, 'ADDRESS')]}), 
                ('send to: Albisstrasse 8038', {'entities': [(9, 25, 'ADDRESS')]}), 
                ('send to: Albulastrasse 8048', {'entities': [(9, 26, 'ADDRESS')]}), 
                ('send to: Alderstrasse 8008', {'entities': [(9, 25, 'ADDRESS')]})]
```
Is it possible to get good results?
I mean, street names can have 1-4 words in them, they can have numbers etc.
Is it possible to get good results because Ill always get different street addresses?

**Comment by polm** [comment]
Whether you can get good results or not depends on what kind of situation you want to recognize street addresses in. 

Are you recognizing streets addresses written in isolation? Or do you want to parse a whole address, with street, city, country, and so on, and isolate the street part? In that case the pretrained spaCy models won't help you much because they're trained on text like newspaper articles that uses complete sentences. But you can train a custom model that can learn how commas and words like "City" or "St." are significant. 

On the other hand, if you're recognizing addresses in sentences ("Mr. Smith lives at 2 3/8 Strawberry Lane..."), the existing models can be a good starting point. 

Either way, your current training data is unhelpful - the point of training data is to show the kind of things you want to find in **in a wide variety of situations**, so repeating the same template over and over again won't work.

What kind of text do you want to use your address matcher on? Do you want to parse the address into parts or just recognize it?

**Reply by aleksandar-devedzic** [reply]
I want to make model for invoice processing. 
So on each invoice Ill have different name, address, company name etc.
Does that mean that I need supply my model with full invoice text ? For example:

` training_data = [('full invoice text with all entities that I want, {'entities': [(9, 28, 'ADDRESS')]}), `

I do not have 1000 invoice examples, I only have 15 examples, so I the previous example I have just added "send to: address"  just as a random text, I taught that model will focus on part of text with address. But can I supply my model with some other data that have company names, addresses etc, and to use that model on invoices?

**Reply by polm** [reply]
OK, thanks for clarifying you want to work on invoices. What does the full invoice text look like though? In invoices I deal with the address is just by itself, floating on the page somewhere, and not in the middle of a sentence. In a context like that it may not be impossible to train a model from just text but it'll certainly be challenging. 

If you only have 15 examples you can't train a model in any case. You should look at using [rule-based matching](https://spacy.io/usage/rule-based-matching), which you can use to get started and create more training data. This is a place that you can use the company names and addresses you have.

**Reply by aleksandar-devedzic** [reply]
My invoices are in PDF file. Im using Python's `pdftotext` lib for extracting data. My texts are between 100-1000 words.

This thing is not clear to me:
Imagine if I had 5000 invoice texts. Those 5000 texts have 5000 addresses, 5000 company names etc.

And data that im passing for training should look like:

`training_data = [('full invoice text that can be long', {'entities': [(start index of address, end index of address, 'ADDRESS')]})
`
How can I know what addresses are in my invoices, how can I know their names, and their start indexes and end indexes, do I need to go over 5000 of my invoices manually and to add them in training_data?
Is there a way for me to download, (for example), 10000 addresses from the internet, and to pass them to training?
Those addresses wont be addresses from my invoices, so I can not know the start index and end index.

**Reply by polm** [reply]
Ideally you should manually create training data, yes. That means going over your documents and marking where the addresses are. 

If you have a good tool like [Prodigy](https://prodi.gy/), doing 1000 examples or so really does not take a long time. 

If you do not want to manually label data, you can do what's called **weak supervision** - you can use rules to create annotations and train a model and improve it gradually over time. For example, with addresses, you could use rules to mark "[anything] St." as an address. This will create false positives, for example you could have "St. John's School" or something, but it's a starting point.

**Reply by aleksandar-devedzic** [reply]
Is Prodigy free tool?
If not, is there a trial version?

**Reply by polm** [reply]
Prodigy is not free. You can check out the [online demo](https://prodi.gy/demo) to get an idea how the UI works, or see the [ordering FAQ](https://prodi.gy/docs/faq#ordering) for information about a trial.

**Reply by aleksandar-devedzic** [reply]
do you think that I can fix this kind of problem with rule-based-matching?

Please check out the question:
https://stackoverflow.com/questions/66821133/creating-rule-based-matching-with-spacy-and-python-for-detecting-addresses

**Reply by polm** [reply]
Yes, I think rule-based matching is a good place to start for your problem. I answered the question on Stack Overflow, but please also read the docs, they explain how to solve problems like this.

Also please do not ping me on StackOverflow as well as here, the extra notifications are not helpful.
